library/Core/src/Juvix/Core/Unify/Types.hs

```diff
@@ -0,0 +1,230 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+
+module Juvix.Core.Unify.Types
+  ( module Juvix.Core.Unify.Term,
+
+    -- * Metavariables
+    MetaVar,
+    MetaSet,
+    MetaMap,
+    metasT,
+    metasE,
+    metasV,
+    metasN,
+    occursT,
+    occursE,
+    occursV,
+    occursN,
+
+    -- * Substitutions
+    Subst (..),
+    appSubstV,
+    appSubstN,
+
+    -- * Unification
+    Unify (..),
+    HasError,
+    HasSubst,
+    HasProblems,
+    HasGlobals,
+    CanUnify,
+    PrimUnify,
+    PrimUnify1,
+    Problem,
+    Error (..),
+    runUnify,
+    throwU,
+    bind,
+    occursCheck,
+    nextProblem,
+    addProblem,
+    addProblems,
+
+    -- * Reexports
+    Universe,
+    GlobalName,
+    PatternVar,
+    PatternMap,
+    BoundVar,
+    Name (..),
+  )
+where
+
+import Data.Sequence (Seq (..))
+import Juvix.Core.IR.Evaluator.Weak
+import qualified Juvix.Core.IR.Types as IR
+import Juvix.Core.IR.Types.Base
+import Juvix.Core.Unify.MetaVar (MetaMap, MetaSet, MetaVar)
+import qualified Juvix.Core.Unify.MetaVar as Meta
+import Juvix.Core.Unify.Subst
+import Juvix.Core.Unify.Term
+import Juvix.Library
+import qualified Juvix.Library.Usage as Usage
+
+metasT :: Term primTy primVal -> MetaSet
+metasT = \case
+  Star _ℓ -> mempty
+  PrimTy _t -> mempty
+  Prim _p -> mempty
+  Pi _π a b -> metasT a <> metasT b
+  Lam t -> metasT t
+  Sig _π a b -> metasT a <> metasT b
+  Pair a b -> metasT a <> metasT b
+  Let _π e b -> metasE e <> metasT b
+  UnitTy -> mempty
+  Unit -> mempty
+  Elim e -> metasE e
+  Meta α -> Meta.singleS α
+
+metasE :: Elim primTy primVal -> MetaSet
+metasE = \case
+  Bound _i -> mempty
+  Free _x -> mempty
+  App f s -> metasE f <> metasT s
+  Ann _π t a _ℓ -> metasT t <> metasT a
+
+metasV :: Value primTy primVal -> MetaSet
+metasV = \case
+  VStar _ℓ -> mempty
+  VPrimTy _t -> mempty
+  VPrim _p -> mempty
+  VPi _π a b -> metasV a <> metasV b
+  VLam t -> metasV t
+  VSig _π a b -> metasV a <> metasV b
+  VPair a b -> metasV a <> metasV b
+  VUnitTy -> mempty
+  VUnit -> mempty
+  VNeutral n -> metasN n
+  VMeta α -> Meta.singleS α
+
+metasN :: Neutral primTy primVal -> MetaSet
+metasN = \case
+  NBound _i -> mempty
+  NFree _x -> mempty
+  NApp f s -> metasN f <> metasV s
+
+occursT :: MetaVar -> Term primTy primVal -> Bool
+occursT α t = α `Meta.memberS` metasT t
+
+occursE :: MetaVar -> Elim primTy primVal -> Bool
+occursE α e = α `Meta.memberS` metasE e
+
+occursV :: MetaVar -> Value primTy primVal -> Bool
+occursV α t = α `Meta.memberS` metasV t
+
+occursN :: MetaVar -> Neutral primTy primVal -> Bool
+occursN α t = α `Meta.memberS` metasN t
+
```

Comment:
A lot of these just seem to repeat themselves due to a very slightly differen representation